### DISCUSSION TITLE
same day event being removed before it occurs

### DISCUSSION THREAD

**Discussion by mschreiberjr** [derekantrican/GAS-ICS-Sync#382] [General]
Hi,  i have events being removed just a few hours before they actually occur.     

Here are my settings.  I have only future events set to true, and remove events not found to true.  

I dont want to sync previous events.  Any way I can sync and have them not be removed until they have passed? 

var howFrequent = 10;                  // What interval (minutes) to run this script on to check for new events
var onlyFutureEvents = true;          // If you turn this to "true", past events will not be synced (this will also removed past events from the target calendar if removeEventsFromCalendar is true)
var addEventsToCalendar = true;        // If you turn this to "false", you can check the log (View > Logs) to make sure your events are being read correctly before turning this on
var modifyExistingEvents = true;       // If you turn this to "false", any event in the feed that was modified after being added to the calendar will not update
var removeEventsFromCalendar = true;   // If you turn this to "true", any event created by the script that is not found in the feed will be removed.
var addAlerts = "yes";                 // Whether to add the ics/ical alerts as notifications on the Google Calendar events or revert to the calendar's default reminders ("yes", "no", "default").
var addOrganizerToTitle = false;       // Whether to prefix the event name with the event organiser for further clarity
var descriptionAsTitles = false;       // Whether to use the ics/ical descriptions as titles (true) or to use the normal titles as titles (false)
var addCalToTitle = false;             // Whether to add the source calendar to title
var addAttendees = false;              // Whether to add the attendee list. If true, duplicate events will be automatically added to the attendees' calendar.
var defaultAllDayReminder = -1

**Comment by Lonestarjeepin** [comment]
I'm not sure why events would be getting deleted before the start time, but I didn't like how events were getting deleted after start time but before end time (e.g. while the event was in progress) when onlyFutureEvents=true.  You could try using #217 and set it to 1 day or something short duration like that so you are only syncing a limited set of historical events, but they don't get deleted in "real time".

**Comment by cgerlfunkenwerfer** [comment]
We are encountering the same issue: Events are deleted from the Calendar some hours before they actually start. Very hard to reproduce, as it seems to happen randomly and it's nowhere to be found in the Logs that something actually gets deleted - the Events can be found in the Google Calendar trash afterwards. 

We just have one execution saying: "Fetched 104 existing events from Calendar", and then the next one saying "Fetched 103 existing events from Calendar", but none of these Executions log ""Deleting old event ".

Here our config:
var howFrequent = 5;                     // What interval (minutes) to run this script on to check for new events
var onlyFutureEvents = true;             // If you turn this to "true", past events will not be synced (this will also removed past events from the target calendar if removeEventsFromCalendar is true)
var addEventsToCalendar = true;           // If you turn this to "false", you can check the log (View > Logs) to make sure your events are being read correctly before turning this on
var modifyExistingEvents = false;          // If you turn this to "false", any event in the feed that was modified after being added to the calendar will not update
var removeEventsFromCalendar = true;      // If you turn this to "true", any event created by the script that is not found in the feed will be removed.
var removePastEventsFromCalendar = false;  // If you turn this to "false", any event that is in the past will not be removed.
var addAlerts = "yes";                    // Whether to add the ics/ical alerts as notifications on the Google Calendar events or revert to the calendar's default reminders ("yes", "no", "default").
var addOrganizerToTitle = false;          // Whether to prefix the event name with the event organiser for further clarity
var descriptionAsTitles = false;          // Whether to use the ics/ical descriptions as titles (true) or to use the normal titles as titles (false)
var addCalToTitle = false;                // Whether to add the source calendar to title
var addAttendees = false;                 // Whether to add the attendee list. If true, duplicate events will be automatically added to the attendees' calendar.
var defaultAllDayReminder = -1;           // Default reminder for all day events in minutes before the day of the event (-1 = no reminder, the value has to be between 0 and 40320)
                                          // See https://github.com/derekantrican/GAS-ICS-Sync/issues/75 for why this is neccessary.
var overrideVisibility = "";              // Changes the visibility of the event ("default", "public", "private", "confidential"). Anything else will revert to the class value of the ICAL event.
var addTasks = false;

Any Idea why this could happen? 
Version is 5.7

**Reply by jonas0b1011001** [reply]
@cgerlfunkenwerfer @mschreiberjr 
In case the first event of a recurring series is affected by onlyFutureEvents the event does not get deleted (as it would remove all future instances aswell) but moved to the next occurence. That's the only case i can think of that an event 'disappears' without a "Deleting old event" log entry.

> We just have one execution saying: "Fetched 104 existing events from Calendar", and then the next one saying "Fetched 103 existing events from Calendar", but none of these Executions log ""Deleting old event ".

Can you check that second log for this entry: 
https://github.com/derekantrican/GAS-ICS-Sync/blob/721716039b473929861139de557277571b7af9d6/Helpers.gs#L637

**Reply by cgerlfunkenwerfer** [reply]
I did some more debugging and found out the following: We are using Timezone Europe/Berlin (UTC+1 currently).
As soon as it's less then 1h before the Event starts, it is no longer present in the "icsEventsIds" Array, when the function processEventCleanup is called. Even though, in the original ics, the event is still present. I think it's a timezone issue?

I have to correct myself from the previous post: The following line is there in the debug log as soon as it's deleted: "Deleting old event scheduler/1906513"
Regarding your comment, it's a single event, non-recurring. And the line you mentioned is not in the logs.

I continuing to investigate, why the event does not make it into the "icsEventsIds" array.

**Reply by jonas0b1011001** [reply]
Does the *.ics contain a VTIMEZONE element for "Europe/Berlin"?
You can check by downloading the *.ics, open in a text editor and search for the "VTIMEZONE" tag. Should be at the beginning of the file.

**Reply by cgerlfunkenwerfer** [reply]
Yes, it does include the time zone: 
Here's the ics, removed all other events than the test event that was just removed from the calendar, 1h before it started:

[cal.txt](https://github.com/derekantrican/GAS-ICS-Sync/files/15151554/cal.txt)

**Reply by cgerlfunkenwerfer** [reply]
So I'm one step further in investigation, it is currently filtered out in this part of the code: 

https://github.com/derekantrican/GAS-ICS-Sync/blob/721716039b473929861139de557277571b7af9d6/Helpers.gs#L229-L243

**Reply by jonas0b1011001** [reply]
I can't really reproduce the 1h threshold but 2 hrs (Europe/Berlin is UTC+2 currently).

Can you check if changing
https://github.com/derekantrican/GAS-ICS-Sync/blob/721716039b473929861139de557277571b7af9d6/Code.gs#L157
to
`startUpdateTime = new ICAL.Time.fromJSDate(new Date(), true);`
changes the behavior.

**Reply by cgerlfunkenwerfer** [reply]
Yes, that fixed it! It is no longer removed. 

Thanks!
I think that should probably be fixed in the code for other users.

**Reply by jonas0b1011001** [reply]
Definitely!

One last request: What timezone is set in your scripts settings? Left sidebar -> "Projekteinstellungen"

**Reply by cgerlfunkenwerfer** [reply]
That was GMT+2:
![Screenshot 2024-04-29 182212](https://github.com/derekantrican/GAS-ICS-Sync/assets/148237190/0372323d-9855-42d0-9143-c4f43c676f46)

**Comment by jonas0b1011001** [comment]
This should be fixed as per https://github.com/derekantrican/GAS-ICS-Sync/pull/437
